hummingbot/strategy/avellaneda_market_making/avellaneda_market_making_config_map_pydantic.py

```diff
@@ -0,0 +1,529 @@
+
+
+class SingleOrderLevelModel(BaseClientModel):
+    class Config:
+        title = OrderLevelsMode.single_order_level
+        validate_assignment = True
+
+
+class MultiOrderLevelModel(BaseClientModel):
+    order_levels: int = Field(
+        default=2,
+        description="The number of orders placed on either side of the order book.",
+        ge=2,
+        client_data=ClientFieldData(
+            prompt=lambda mi: "How many orders do you want to place on both sides?"
+        ),
+    )
+    level_distances: Decimal = Field(
+        default=Decimal("0"),
+        description="The spread between order levels, expressed in % of optimal spread.",
+        ge=0,
+        client_data=ClientFieldData(
+            prompt=lambda mi: "How far apart in % of optimal spread should orders on one side be?"
+        ),
+    )
+
+    class Config:
+        title = OrderLevelsMode.multi_order_level
+        validate_assignment = True
+
+    @validator("order_levels", pre=True)
+    def validate_int_zero_or_above(cls, v: str):
+        ret = validate_int(v, min_value=2)
+        if ret is not None:
+            raise ValueError(ret)
+        return v
+
+    @validator("level_distances", pre=True)
+    def validate_decimal_zero_or_above(cls, v: str):
+        ret = validate_decimal(v, min_value=Decimal("0"), inclusive=True)
+        if ret is not None:
+            raise ValueError(ret)
+        return v
+
+
+class HangingOrdersMode(str, ClientConfigEnum):
+    track_hanging_orders = "track_hanging_orders"
+    ignore_hanging_orders = "ignore_hanging_orders"
+
+
+class TrackHangingOrdersModel(BaseClientModel):
+    hanging_orders_cancel_pct: Decimal = Field(
+        default=Decimal("10"),
+        description="The spread percentage at which hanging orders will be cancelled.",
+        gt=0,
+        lt=100,
+        client_data=ClientFieldData(
+            prompt=lambda mi: (
+                "At what spread percentage (from mid price) will hanging orders be canceled?"
+                " (Enter 1 to indicate 1%)"
+            ),
+        )
+    )
+
+    class Config:
+        title = HangingOrdersMode.track_hanging_orders
+        validate_assignment = True
+
+    @validator("hanging_orders_cancel_pct", pre=True)
+    def validate_pct_exclusive(cls, v: str):
+        ret = validate_decimal(v, min_value=Decimal("0"), max_value=Decimal("100"), inclusive=False)
+        if ret is not None:
+            raise ValueError(ret)
+        return v
+
+
+class IgnoreHangingOrdersModel(BaseClientModel):
+    class Config:
+        title = HangingOrdersMode.ignore_hanging_orders
+        validate_assignment = True
+
+
+def maker_trading_pair_prompt(model_instance: 'AvellanedaMarketMakingConfigMap') -> str:
+    exchange = model_instance.exchange
+    example = AllConnectorSettings.get_example_pairs().get(exchange)
+    return (
+        f"Enter the token trading pair you would like to trade on {exchange}{f' (e.g. {example})' if example else ''}"
+    )
```

Comment:
Would it be possible to make prompt functions either class or static functions in each conf model class?

Reply:
A couple of issues here.

The first one is that they would have to be defined before the usage, so 
```Python
class AvellanedaMarketMakingConfigMap(BaseClientModel):
    @staticmethod
    def maker_trading_pair_prompt(model_instance: 'AvellanedaMarketMakingConfigMap') -> str:
        exchange = model_instance.exchange
        example = AllConnectorSettings.get_example_pairs().get(exchange)
        return (
            f"Enter the token trading pair you would like to trade on {exchange}{f' (e.g. {example})' if example else ''}"
        )

    market: str = Field(
        default=...,
        description="The trading pair.",
        client_data=ClientFieldData(
            prompt=maker_trading_pair_prompt,
            prompt_on_new=True,
        ),
    )
```
which is ok.

The second issue is that the above approach raises `TypeError: 'staticmethod' object is not callable`. There are two solutions proposed [here](https://stackoverflow.com/questions/41921255/staticmethod-object-is-not-callable). Either do
```Python
    market: str = Field(
        default=...,
        description="The trading pair.",
        client_data=ClientFieldData(
            prompt=maker_trading_pair_prompt.__get__(object),
            prompt_on_new=True,
        ),
    )
```
or
```Python
class AvellanedaMarketMakingConfigMap(BaseClientModel):
    def maker_trading_pair_prompt(model_instance: 'AvellanedaMarketMakingConfigMap') -> str:
        exchange = model_instance.exchange
        example = AllConnectorSettings.get_example_pairs().get(exchange)
        return (
            f"Enter the token trading pair you would like to trade on {exchange}{f' (e.g. {example})' if example else ''}"
        )
```
but I think both will be confusing for users (and for myself when I read the code two months from now)

Reply:
What happens if the function is a class function? Can we configure it as `prompt=AvellanedaMarketMakingConfigMap.maker_trading_pair_prompt` ?

Reply:
Unfortunately no, because at that time the class is not yet defined

Reply:
One final question: since the prompt can be assigned any callable, could we assign it a lambda that calls the class method maybe? That could help us keep it all packed together.

Reply:
Nice. That did the trick